cache/remotecache/s3/s3_downloader.go

```diff
@@ -0,0 +1,219 @@
+package s3
+
+// To be replaced by direct SDK call when available
+// https://github.com/aws/aws-sdk-go-v2/issues/2247
+
+import (
+	"context"
+	"io"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/containerd/containerd/content"
+	"golang.org/x/sync/errgroup"
+)
+
+type S3Downloader struct {
+	input *S3DownloaderInput
+}
+
+type S3DownloaderInput struct {
+	Bucket      *string
+	Key         *string
+	S3Client    *s3.Client
+	Size        int64
+	Parallelism int
+	PartSize    int
+}
+
+type chunkStatus struct {
+	writeOffset int64
+	readOffset  int64
+	done        bool
+	size        int64
+	start       int64
+	buffer      []byte
+	io.Writer
+}
+
+func (chunk *chunkStatus) Write(p []byte) (n int, err error) {
+	copied := copy(chunk.buffer[chunk.writeOffset:], p)
+	chunk.writeOffset += int64(copied)
+	return copied, nil
+}
+
+type s3Reader struct {
+	content.ReaderAt
+	inChan        chan int
+	outChan       chan error
+	start         int64
+	currentOffset int64
+	currentChunk  int
+	input         *S3DownloaderInput
+	ctx           context.Context
+	eg            *errgroup.Group
+	groupCtx      context.Context
+	totalChunk    int
+	chunks        []chunkStatus
+}
+
+func newDownloader(input *S3DownloaderInput) *S3Downloader {
+	if input.Parallelism == 0 {
+		input.Parallelism = 8
```

Comment:
You can test this, but for registry layers, our default is 4.

Reply:
These default are not really used, they are overriden in s3.go, where the default value is 4. I changed them anyway.